src/ExtMsg.actor.cpp

```diff
@@ -626,41 +626,36 @@ ACTOR Future<Reference<IReadWriteContext>> insertDocument(Reference<CollectionCo
                                                           bson::BSONObj d,
                                                           Optional<IdInfo> encodedIds) {
 	state Standalone<StringRef> encodedId;
-	state Standalone<StringRef> valueEncodedId;
 	state Optional<bson::BSONObj> idObj;
 	state Reference<QueryContext> dcx;
 
 	if (!encodedIds.present()) {
-		encodedId = DataValue(bson::OID::gen()).encode_key_part();
-		valueEncodedId = encodedId;
-		idObj = Optional<bson::BSONObj>();
+		// No _id field, generate a new one.
+		state bson::OID generatedId = bson::OID::gen();
+		state bson::BSONObj docWithGeneratedId =
+		    bson::BSONObjBuilder().appendElements(d).append("_id", generatedId).obj();
```

Comment:
Before we were inserting key encoded object as the value, now we are inserting unencoded one. For what its worth key encoding is NOP for OID type.

Reply:
Yeah. Also later in the `insertElementRecursive` the filed name and value will be encoded accordingly.